pppd/main.c

```diff
@@ -747,6 +750,108 @@ set_ifunit(int iskey)
     }
 }
 
+/*
+ * net-* scripts to be run come through here.
+ */
+static void main_script(char* script, int wait)
+{
+    char strspeed[32];
+    char *argv[6];
+
+    slprintf(strspeed, sizeof(strspeed), "%d", baud_rate);
+
+    argv[0] = script;
+    argv[1] = ifname;
+    argv[2] = devnam;
+    argv[3] = strspeed;
+    argv[4] = ipparam;
+    argv[5] = NULL;
+
+    run_program(script, argv, 0, NULL, NULL, wait);
+}
+
+/*
+ * set_ifup - called in order to set the ppp interface to up, if not already
+ * brought up.
+ *
+ * It's important that the sub-protocol not be a dynamically allocated string, ie,
+ * use set_ipup("IPCP") ...
+ */
+int
+set_ifup(const char* name)
+{
+    int i;
+    const char** t;
+
+    if (!up_protos) {
+	main_script(PPP_PATH_NET_PREUP, 1);
+	if (!netif_set_up())
+	    return 0;
+    }
+
+    i = 0;
+    while (up_protos && up_protos[i]) {
+	if (up_protos[i] == name) {
+	    warn("%s previously signalled ifup, this is probably a bug in the protocol.",
+		    name);
+	    return 1;
+	}
+	++i;
+    }
+    info("%s is now UP", name);
+
+    /* i is index where we insert the sub protocol */
+    t = realloc(up_protos, (i + 2) * sizeof(*up_protos));
+    if (t == NULL) {
+	error("Memory allocation error trying to add %s to the list of UP protocols.", name);
+    } else {
+	up_protos = t;
+	up_protos[i] = name;
+	up_protos[i+1] = NULL;
+    }
+
+    return 1;
+}
+
+/*
+ * set_ifdown - called in order to set the ppp interface to down, the interface will
+ * only be downned if all sub protocols have signalled down.
+ *
+ * It's important that the sub-protocol not be a dynamically allocated string, ie,
+ * use set_ipup("IPCP") ...
+ */
+int
+set_ifdown(const char* name)
+{
+    int i = 0;
+    const char** t;
+
+    while (up_protos && up_protos[i] && up_protos[i] != name)
```

Comment:
Why not use a strcmp() here instead of comparing the memory address?

Reply:
no real reason other than a compiler shouldn't duplicate the string and as such a pointer comparison is good enough.  I've never seen that it gets duplicated but I've only ever worked with gcc in the last 19 years and some MSVC prior to that.  Can convert if that's preferred, doesn't make a significant difference, and since others will probably wonder the same thing it's probably worth the change from a sanity perspective and just in case some compiler does actually duplicate the strings?